plugins/inputs/open_hardware_monitor/open_hardware_monitor.go

```diff
@@ -1,114 +1,198 @@
-			dst[i].Parent = strings.Replace(strings.Trim(dst[i].Parent, key), key, value, -1)
-		}
+
+	return query
+}
+
+func (p *Config) QueryHardware() ([]Hardware, error) {
+	hardwareQuery := p.CreateHardwareQuery()
+
+	var hardware []Hardware
+	err := wmi.QueryNamespace(hardwareQuery, &hardware, "root/OpenHardwareMonitor")
+
+	return hardware, err
+}
+
+func (p *Config) CreateSensorsQuery() string {
+	query := "SELECT * FROM SENSOR"
+
+	sensorConditions := BuildOrConditions(p.Sensor, "Identifier")
+	sensorTypeConditions := BuildOrConditions(p.SensorType, "SensorType")
+	hardwareConditions := BuildOrConditions(p.Hardware, "Parent")
+	hardwareTypeConditions := BuildOrConditions(p.HardwareType, "HardwareType")
+
+	allConditions := BuildAndConditions([]string{
+		sensorConditions,
+		sensorTypeConditions,
+		hardwareConditions,
+		hardwareTypeConditions,
+	})
+
+	if len(allConditions) != 0 {
+		query += " WHERE " + allConditions
 	}
 
-	return dst, err
+	return query
 }
 
-func contains(s []string, e string) bool {
-	if len(s) == 0 {
-		return true
+func (p *Config) QuerySensors() ([]Sensor, error) {
+	sensorsQuery := p.CreateSensorsQuery()
+
+	var sensors []Sensor
+	err := wmi.QueryNamespace(sensorsQuery, &sensors, "root/OpenHardwareMonitor")
+
+	return sensors, err
+}
+
+func BuildTelegrafData(sensor Sensor, hardware Hardware) (map[string]interface{}, map[string]string) {
+	// Field is just the sensor value
+	fields := map[string]interface{}{
+		sensor.SensorType: sensor.Value,
 	}
-	for _, a := range s {
-		if a == e {
-			return true
-		}
+
+	tags := map[string]string{
+		// Sensor info
+		"Sensor_Identifier": sensor.Identifier,
+		"Sensor_Index": strconv.Itoa(sensor.Index),
+		"Sensor_InstanceId": sensor.InstanceId,
+		"Sensor_Name": sensor.Name,
+		"Sensor_Parent": sensor.Parent,
+
+		// Hardware info
+		"Hardware_HardwareType": hardware.HardwareType,
+		"Hardware_Identifier": hardware.Identifier,
+		"Hardware_InstanceId": hardware.InstanceId,
+		"Hardware_Name": hardware.Name,
 	}
-	return false
+
+	return fields, tags
 }
 
-func (p *OpenHardwareMonitorConfig) Gather(acc telegraf.Accumulator) error {
-	var dst []OpenHardwareMonitorData
-	query, err := p.CreateQuery()
-	if err == nil {
-		dst, err = p.QueryData(query)
-		if err != nil {
-			acc.AddError(err)
-		}
-		for _, sensorData := range dst {
-			if contains(p.Parent, sensorData.Parent) {
-				tags := map[string]string{
-					"name":   sensorData.Name,
-					"parent": sensorData.Parent,
-				}
-				fields := map[string]interface{}{sensorData.SensorType: sensorData.Value}
-				acc.AddFields("ohm", fields, tags)
-			}
-		}
-	} else {
+func (p *Config) Gather(acc telegraf.Accumulator) error {
+	var hardware []Hardware
+	hardware, err := p.QueryHardware()
+
+	if err != nil {
 		acc.AddError(err)
 	}
+
+	hardwareByIdentifier := map[string]Hardware{}
+	for _, h := range hardware {
+		hardwareByIdentifier[h.Identifier] = h
+	}
+
+	var sensors []Sensor
+	sensors, err = p.QuerySensors()
```

Comment:
Unfortunately it does not work well. 
Using configuration file: 
```
 [[inputs.open_hardware_monitor]]
 	HardwareType = ["GpuNvidia"] # optional
```

You are building sensor query using attribute "HardwareType" which is not available in Sensor query class: 

It result with query error: 
```
PS D:\Temp\ohm> .\open_hardware_monitor-telegraf-plugin.exe --config .\open_hardware_monitor-telegraf-plugin.config
2022/09/05 12:24:12 E! Error in plugin: Exception occurred. (Invalid query )
```

Analysing the source code, the wmi query looks like this: 
```
Get-WmiObject -Namespace root/OpenHardwareMonitor -Query "SELECT * FROM SENSOR WHERE (HardwareType='GpuNvidia')"
```
which lead to an query error: 

```
Get-WmiObject : Invalid query "SELECT * FROM SENSOR WHERE (HardwareType='GpuNvidia')"
At line:1 char:1
+ Get-WmiObject -Namespace root/OpenHardwareMonitor -Query "SELECT * FR ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Get-WmiObject], ManagementException
    + FullyQualifiedErrorId : GetWMIManagementException,Microsoft.PowerShell.Commands.GetWmiObjectCommand
```

Sensor response for command: 
`
Get-WmiObject -Namespace root/OpenHardwareMonitor -Query "SELECT * FROM SENSOR"
`
returns: 
```
__GENUS          : 2
__CLASS          : Sensor
__SUPERCLASS     :
__DYNASTY        : Sensor
__RELPATH        : Sensor.InstanceId="3911",ProcessId="67ccdffb-07e7-45fc-bdb6-f017483b0a9f"
__PROPERTY_COUNT : 10
__DERIVATION     : {}
__SERVER         : MARIANO-PC
__NAMESPACE      : root\OpenHardwareMonitor
__PATH           : \\MARIANO-PC\root\OpenHardwareMonitor:Sensor.InstanceId="3911",ProcessId="67ccdffb-07e7-45fc-bdb6-f017483b0a9f"
Identifier       : /lpc/nct6798d/fan/1
Index            : 1
InstanceId       : 3911
Max              : 1032.11
Min              : 601.0685
Name             : Fan #2
Parent           : /lpc/nct6798d
ProcessId        : 67ccdffb-07e7-45fc-bdb6-f017483b0a9f
SensorType       : Fan
Value            : 611.6901
PSComputerName   : MARIANO-PC
```
So there is no "HardwareType" field.